vllm/entrypoints/openai/tools.py

```diff
@@ -0,0 +1,201 @@
+
+    def render_toolmessage(self, message: ChatCompletionToolMessage):
+        return self.template.render(
+            CONTEXT=ToolsCallsTemplateContext.TOOL_RESPONSE, message=message)
+
+    def render_toolslist(self, tool_choice: Union[str, None],
+                         tools_list: [ChatCompletionToolParam]) -> str:
+        if isinstance(tool_choice, str) and tool_choice == "auto":
+            tool_choice = None
+        if tool_choice is not None:
+            for tool in tools_list:
+                # Search if the tool_choice is in the tools_list
+                if tool.type == "function" and tool.function.name == tool_choice:
+                    return self.template.render(
+                        CONTEXT=ToolsCallsTemplateContext.FORCE_CALL,
+                        tool=tool)
+            return None
+        else:
+            return self.template.render(
+                CONTEXT=ToolsCallsTemplateContext.FUNCTIONS_LIST,
+                tools_list=tools_list)
+
+
+class OpenAIToolsPrompter:
+    """
+    https://platform.openai.com/docs/assistants/tools
+    """
+
+    def __init__(self, template_path=None):
+        self.template = ToolsCallsTemplate(template_path)
+        self.call_token_str = self.template.get_func_call_token()
+        if self.call_token_str is None:
+            logger.error("There is something wrong with the tools template.")
+        else:
+            self.call_token_pre = self.call_token_str[0]
+
+    def func_call_token_pre(self) -> str:
+        return self.call_token_pre
+
+    def func_call_token_size(self) -> int:
+        return len(self.call_token_str)
+
+    def func_call_token(self) -> str:
+        return self.call_token_str
+
+    def content_from_assistant(self,
+                               message: ChatCompletionAssistantMessage) -> str:
+        text = self.template.render_toolcalls(message.tool_calls)
+        if message.content is None:
+            return text
+        else:
+            return message.content + "\n" + text
+
+    def content_from_tool(self, message: ChatCompletionToolMessage) -> str:
+        return self.template.render_toolmessage(message)
+
+    def inject_prompt(self, request: ChatCompletionRequest):
+        """ Generate and inject the prompt for tools calls. """
+        if request.tools is not None and self.call_token_str is not None and len(
+                request.tools):
+            select_tool_choice = request.tool_choice if (
+                request.tool_choice is not None
+                and request.tool_choice != "auto") else None
+            text_inject = self.template.render_toolslist(
+                tool_choice=select_tool_choice, tools_list=request.tools)
+            if isinstance(request.messages, str):
+                request.messages = text_inject + request.messages
+            elif isinstance(request.messages,
+                            List) and len(request.messages) >= 1:
+                request.messages[
```

Comment:
I think this can be replaced by adding another `system` message.
`request.messages.insert(0,ChatCompletionSystemMessage(role='system', content=text_inject))`

Reply:
I don't thinks this is a good idea. Some models does't have `system` role, plus, some others template (like llama), doesn't allow 2 system messages during the conversation (it will preserve the first system msg and ignore any other).